src/basic_recipes/voronoiplot.jl

```diff
@@ -0,0 +1,228 @@
+- `markercolor = :black` sets the color of the points.
+
+- `strokecolor = :black` sets the strokecolor of the polygons.
+- `strokewidth = 1` sets the width of the polygon stroke.
+- `color = automatic` sets the color of the polygons. If `automatic`, the polygons will be individually colored according to the colormap.
+- `unbounded_edge_extension_factor = 0.1` sets the extension factor for the unbounded edges, used in `DelaunayTriangulation.polygon_bounds`.
+- `bounding_box = automatic` sets the bounding box for the polygons. If `automatic`, the bounding box will be determined automatically based on the extension factor, otherwise it should be a `Tuple` of the form `(xmin, xmax, ymin, ymax)`. If any of the generators or polygons are outside of the polygon, the plot will error.
+
+$(Base.Docs.doc(MakieCore.colormap_attributes!))
+"""
+@recipe(Voronoiplot, vorn) do scene
+    th = default_theme(scene, Mesh)
+    sc = default_theme(scene, Scatter)
+    attr = Attributes(;
+                      # Toggles
+                      show_generators=true,
+                      smooth=false,
+
+                      # Point settings
+                      markersize=4,
+                      marker=sc.marker,
+                      markercolor=sc.color,
+
+                      # Polygon settings
+                      strokecolor=theme(scene, :patchstrokecolor),
+                      strokewidth=1.0,
+                      color=automatic,
+                      unbounded_edge_extension_factor=0.1,
+                      bounding_box=automatic)
+    MakieCore.colormap_attributes!(attr, theme(scene, :colormap))
+    return attr
+end
+
+function _clip_polygon(poly::Polygon, circle::Circle)
+    # Sutherland-Hodgman adjusted
+    @assert isempty(poly.interiors) "Polygon must not have holes for clipping."
+
+    function intersection(A, B, circle)
+        CA = A - origin(circle)
+        AB = B - A
+        a = dot(AB, AB) # > 0
+        b = 2 * dot(CA, AB) # > 0
+        c = dot(CA, CA) - radius(circle) * radius(circle)
+        t = (sqrt(b*b - 4 * a * c) - b) / (2a) # only solution > 0 matters
+        return A + AB * t
+    end
+
+    input = Point2f.(first.(poly.exterior))
+    output = sizehint!(Point2f[], length(input))
+
+    for i in eachindex(input)
+        p1 = input[mod1(i-1, end)]
+        p2 = input[i]
+
+        Cp1 = p1 - origin(circle)
+        Cp2 = p2 - origin(circle)
+        r2 = radius(circle) * radius(circle)
+        if dot(Cp1, Cp1) < r2 # p1 inside
+            if dot(Cp2, Cp2) > r2 # p2 outside
+                p = intersection(p1, p2, circle)
+                push!(output, p)
+            else # both inside
+                push!(output, p2)
+            end
+        elseif dot(Cp2, Cp2) < r2 # p1 outside, p2 inside
+            p = intersection(p2, p1, circle)
+            push!(output, p, p2)
+        end
+    end
+
+    return Polygon(output)
+end
+_clip_polygon(poly::Polygon, ::Any) = poly
```

Comment:
I'm not sure this can easily work. The clipped polygons against the circle will not match with the actual circle itself, you would need to place arcs between points on the circle. It also doesn't seem to currently work. 

```julia
A = (-3.0, 7.0)
B = (1.0, 6.0)
C = (-1.0, 3.0)
D = (-2.0, 4.0)
E = (3.0, -2.0)
F = (5.0, 5.0)
G = (-4.0, -3.0)
H = (3.0, 8.0)
points = [A, B, C, D, E, F, G, H]
tri = triangulate(points)
vorn = voronoi(tri)
fig = Figure()
ax = PolarAxis(fig[1, 1])
voronoiplot!(ax, vorn, show_generators = true, markersize = 14, color = :white)
ax = Axis(fig[1, 2]) # just to see what it actually looks like 
voronoiplot!(ax, vorn, show_generators = true, markersize = 14, color = :white)
fig
```

![32918029b800569033bc49e2171bdad2](https://github.com/MakieOrg/Makie.jl/assets/95613936/07c9b680-e617-4337-bed4-ce69389a73e7)

The polygons clip together as you can see, and you can see what I mention about needing arcs: The polygon near 45 and 90 has an empty space on the edge connecting two points on the boundary.

Reply:
Oh, the issue with the first is because the conversion to a circular bounding box only happens when you pass points, not when you pass a tessellation itself. Is that the behaviour you prefer?

Reply:
Passing the circular bounding box itself doesn't seem to be any better actually:

```julia
fig = Figure()
ax = PolarAxis(fig[1, 1])
c = Makie.Circle(Makie.Point2f(0.0, 0.0), 15.0)
voronoiplot!(ax, vorn, show_generators = true, markersize = 14, color = :white, bounding_box = c)
ax = Axis(fig[1, 2]) # just to see what it actually looks like 
voronoiplot!(ax, vorn, show_generators = true, markersize = 14, color = :white)
fig
```

![5f637cc088d03f98c9f404629133eb48](https://github.com/MakieOrg/Makie.jl/assets/95613936/a604d89d-2eab-4a2e-8bb5-919787d000fe)


But if it's intentional that this should be only when `points` are provided (which I've not tested) then I guess this is fine.

Reply:
There is a bit of a messy situation here with transform functions. If you do
```julia
ax = PolarAxis(fig[1, 1])
voronoiplot!(ax, vorn)
```
you end up applying the voronoi tessellation in Polar coordinates, which I assume just doesn't make sense. The Polar -> Cartesian transformation will apply later down the line in poly and scatter.

If you pass points directly we can resolve this in the `plot!` method beforehand, i.e. do Polar -> Cartesian with
```julia
transformed = Makie.apply_transform(tf, ps)
```
and then do the voronoi tessellation in Cartesian space. To avoid applying the Polar -> Cartesian transformation twice we remove it with
```julia
transform = Transformation(p.transformation; transform_func=identity)
```

If we wanted to resolve the polar transform in both cases we would have to extract the input points, transform them and then rerun the tessellation. I'm not sure if that's possible (can we recover how the tessellation was set up?) and it also suggests doing things inefficiently (double tessellation if you're not careful). I'd rather just have the `VoronoiTessellation` method be the "I know what I'm doing" version and the `Vector{Point2f}` method be the "Do what I want" version.

> Passing the circular bounding box itself doesn't seem to be any better actually:

Same problem here - when plotting to a PolarAxis and directly passing a tessellation the bounding box applies in polar coordinates.

Reply:
That makes sense. Would it be sensible to disallow passing `vorn` like this into a `PolarAxis` then? Or just trust the user to "know what they're doing" as you say? I'm up for either.

Reply:
> I'm not sure this can easily work. The clipped polygons against the circle will not match with the actual circle itself, you would need to place arcs between points on the circle.

You can make the same argument for the polygons in general. If the boundary is curved, why not all the edges? To me it looks kind of weird if it's just the outer ones: (quick test)
![Screenshot 2023-08-12 015801](https://github.com/MakieOrg/Makie.jl/assets/10947937/4716e9a8-4308-44d1-8aaa-e7d47d573a3b)
I think in the end this will just result in dragging out the pr a lot for relatively little gain.

Reply:
Fair point. I'm not too fussed either way, just wanted to make sure you are aware. Let us not worry about it :)